### PR TITLE
fix(sampler): decouple greedy and topk-topp sampling

### PR DESCRIPTION
**Pull request by rebel-eunji**
## AS-IS

- If top_k and top_p are both None (all greedy), use greedy sampling
- Greedy case also uses the topk_topp torch op

## TO-BE

- If top_k and top_p are both None, use random sampling
- Greedy case uses a separate greedy (argmax) torch op

## NOTE
- Handling for the case where top_k/top_p are not None but some requests are greedy (greedy + top_k/top_p mixed):
    - Same approach as the rejection sampler — set temperature to an epsilon value so the computation behaves similarly to greedy @rebel-magunudo

> *What does this PR do? What feature, fix, or improvement does it bring?*

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [ ] 🚀 Release (`release`)
* [ ] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [x] 🛠 Bug fix (`fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

1. Run `...`
2. Verify output: `...`
3. Edge case tested: `...`

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [ ] PR title follows Conventional Commits format
* [ ] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

---

### CODE REVIEW

File: vllm_rbln/v1/sample/rbln_sampler.py [hunk 1]
```diff
@@ -329,8 +415,13 @@ def apply_temperature(
...
         # in-place division triggers buffer key error
         # in torchinductor
+        # NOTE:
+        # Greedy requests use a small temperature (1e-3) so softmax collapses
+        # to a near one-hot at argmax. _SAMPLING_EPS (1e-5) is too small here —
+        # it pushes logits past softmax's safe exp range and overflows.
+        _GREEDY_EPS = 1e-3
         if not all_random:
-            temp = torch.where(temp < _SAMPLING_EPS, 1.0, temp)
+            temp = torch.where(temp < _SAMPLING_EPS, _GREEDY_EPS, temp)
```
Comment (rebel-eunji) [inline vllm_rbln/v1/sample/rbln_sampler.py]: Based on a few test cases, _SAMPLING_EPS turned out to be too small considering that logits usually fall in the -20 to +20 range. To address this, I added a separate _GREEDY_EPS with a value of 1e-3. 
@rebel-jongho

Comment (rebel-jongho) [reply vllm_rbln/v1/sample/rbln_sampler.py]: nit: Since `temp` usually stands for 'temporary,' let's spell out `temperature` to keep the code clear

**Review by rebel-jongho — APPROVED** [review]
Just out of curiosity, since the NPU supports greedy sampling, can't we just use the upstream `sample` function as is, not overriding?

File: vllm_rbln/v1/sample/rbln_sampler.py [hunk 2]
```diff
@@ -229,6 +233,88 @@ def __init__(
...
+            sampling_metadata.top_k,
+            sampling_metadata.top_p,
+        )
+
+        assert greedy_sampled is None, (
+            "Upstream vLLM runs greedy and random sampling "
+            "separately and merges the results, "
+            "but vLLM RBLN processes greedy and random requests together: "
+            "greedy requests are routed through the random-sampling path "
+            "with a very small temperature value."
```
Comment (rebel-eunji) [inline vllm_rbln/v1/sample/rbln_sampler.py:315]: @rebel-jongho Upstream vLLM invokes both the greedy and random sampling ops, then merges the results.
### diff
- https://github.com/vllm-project/vllm/blob/694d9a81bbb07977e7a72a597acb44f6a848f774/vllm/v1/sample/sampler.py#L252
   - Upstream vLLM invokes the greedy op whenever not all random (i.e., when at least some requests are greedy), whereas our implementation only invokes it when all_greedy.
- https://github.com/vllm-project/vllm/blob/694d9a81bbb07977e7a72a597acb44f6a848f774/vllm/v1/sample/sampler.py#L288-L296
    - I added assert like this line.

File: vllm_rbln/v1/sample/rbln_sampler.py (hunk 1 above)
Comment (rebel-eunji) [reply vllm_rbln/v1/sample/rbln_sampler.py]: https://github.com/RBLN-SW/vllm-rbln/pull/571/changes/a5a15200db272e325ad40b757d93d36a47023331

**Comment by rebel-eunji** [comment]
The PR CI failed because the latest compiler is not built on server.

File: vllm_rbln/v1/sample/rbln_sampler.py [hunk 3]
```diff
@@ -137,6 +130,19 @@ def rbln_top_k_top_p_sample(
...
+# NOTE(eunji): argmax op is registered in the compiler
+
+
+def rbln_greedy_sample(logits: torch.Tensor) -> torch.Tensor:
+    """
+    Implementation of RBLN greedy sampling.
+    To avoid self parameter issues when torch.compile is used,
+    we define this as a static method.
+    """
+    sampled = torch.ops.rbln.argmax(logits)
```
Comment (huijjj) [inline vllm_rbln/v1/sample/rbln_sampler.py:142]: @rebel-eunji 
It seems like we do not have this yet, would it be possible after the next release?

cc: @pei0033 @junstar92

Comment (rebel-eunji) [reply vllm_rbln/v1/sample/rbln_sampler.py:142]: Sorry for the inconvenience.
This will be available starting from compiler v0.10.4.
Until then, please temporarily replace `torch.ops.rbln.argmax(logits)` with `torch.argmax(dim=-1)`.

Comment (huijjj) [reply vllm_rbln/v1/sample/rbln_sampler.py:142]: Thank you for letting us know
